File: built-in plugins dsp/mb-ethnic-tabla.cpp
/**
 * MB Tabla
 * Category : instrument
 * Type     : ethnic
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Indian tabla drum pair with tuned pitch bends
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_ETHNIC_TABLA_H
#define MB_ETHNIC_TABLA_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbEthnicTabla : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-ethnic-tabla";
    static constexpr const char* PLUGIN_NAME    = "MB Tabla";
    static constexpr const char* PLUGIN_TYPE    = "ethnic";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float tuning = 0.5f;  // range [0, 1]
    float slap = 0.6f;  // range [0, 1]
    float bend = 0.4f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbEthnicTabla() = default;
    ~MbEthnicTabla() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.tuning = std::clamp(params.tuning, 0f, 1f);
        params.slap = std::clamp(params.slap, 0f, 1f);
        params.bend = std::clamp(params.bend, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Tabla
        return input;
    }
};

#endif // MB_ETHNIC_TABLA_H

File: built-in plugins dsp/mb-ethnic-tabla.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbEthnicTablaPlugin: PluginDefinition = {
  id: "mb-ethnic-tabla",
  slug: "mb-ethnic-tabla",
  name: "MB Tabla",
  category: "instrument",
  type: "ethnic" as any,
  version: "1.0.0",
  description: "Indian tabla drum pair with tuned pitch bends",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 0.6 },
    { type: "noise", detune: 0, gain: 0.2 },
    { type: "triangle", detune: 0, gain: 0.2 },
  ],
  envelope: { attack: 0.001, decay: 0.5, sustain: 0.0, release: 0.3 },
  parameters: [
    {
      id: "tuning",
      name: "Tuning",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "slap",
      name: "Slap",
      type: "float",
      defaultValue: 0.6,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "bend",
      name: "Pitch Bend",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { tuning: 0.5, slap: 0.6, bend: 0.4, volume: 0.8 },
};

export default MbEthnicTablaPlugin;
